Add UI for controlling font size and line height settings.

sahil839 opened the pull request:
<!-- Describe your pull request here.-->
- First commit is fix for an existing bug.
- Second commit is to add UI in personal menu.
- Third commit is to add UI in settings UI.
 
Fixes: <!-- Issue link, or clear description.--> #33186

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->
**Screenshots and screen captures:**
| | Screenshots |
| ---------- | ---------- |
| Personal popover light theme | ![popover-info-density-light](https://github.com/user-attachments/assets/9ba5034c-bfe0-434f-907c-98e986f7e760) |
| Personal popover dark theme | ![popover-info-density-dark](https://github.com/user-attachments/assets/87fc8091-a76f-485e-96a2-0ae1aed2dca3) |
| Preferences UI light theme | ![settings-info-density-light](https://github.com/user-attachments/assets/84a4e580-3a61-44a8-9753-58a9f668ecba) |
| Preferences UI dark theme | ![settings-info-density-dark](https://github.com/user-attachments/assets/238f03b0-b5b2-4b4b-b960-f23eec820f33) |
| User defaults UI disabled for non-admins light theme | ![info-density-defaults-non-admins-light](https://github.com/user-attachments/assets/28e1dc59-5674-4da0-a334-f5f81745b019) |
| User defaults UI disabled for non-admins dark theme | ![info-density-defaults-non-admin-dark](https://github.com/user-attachments/assets/b6aca10b-f35e-4d73-864c-83b61d3ef18c) |


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

sahil839 commented:
GIFs
| | GIFs|
| ------- | ------- |
| Personal popover | ![popover-info-density](https://github.com/user-attachments/assets/91c8bd0e-a282-4c39-83cf-90b53124cf3c) |
| Settings UI | ![settings-info-density](https://github.com/user-attachments/assets/2edc930c-6aa2-4959-8b04-058f4e74c54b) |
| User defaults UI | ![settings-info-density-defaults](https://github.com/user-attachments/assets/926c8c3e-7833-44aa-a820-745ee9d5a8a9) |
| Error case when updating through settings UI | ![settings-info-density-error](https://github.com/user-attachments/assets/3b8fb2e4-4f8c-495e-91d4-39a8ec2c7b5b) |

sahil839 commented:
@alya @timabbott Still some minor changes are needed, but I have opened the PR for design and UX feedback and also to get some clarity on a couple of questions.

Things left -
- Need to add attributions for icons.
- Alignment is not perfect for smaller font sizes in settings UI. Working on that.
- There are some possible code deduplications like in templates.

Questions and doubts that I have -
- Am not sure about the colors in disabled state, so would be good to have some feedback on it. And also some suggestions on how to show when individual buttons are disabled.
- Do we want to keep `dense_mode` field now? I think we kept that for easy toggling between 14/122 and 16/140 configurations in prod, but now we provide users a better control.
- I am not sure whether we need tooltips in settings UI as we show the current value there. And especially for line-height, the current value shown are 140, 149, 158 while the tooltip shows 0%, 9%, 18%.

alya commented:
Thanks! The settings UI should use the +x% numbers, not 122, etc.

alya commented:
This is so cool!

Testing the PR:

1. Let's use italics on the current value tooltip lines, as indicated in the issue.
2. For disabled +/- buttons, I think we do want a tooltip after all. Probably a default pointer + "Already at maximum ({value})" / "Already at minimum ({value})" should be OK.
3. For the disabled "reset to default" button, let's also use the default cursor, plus the tooltip specified in the issue.
4. The line spacing tooltips/labels should be, e.g., "+0%", not "0%".
5. The % numbers are wrong. They should all be %s of 140. So, e.g., 131/140 ~= 0.935, so it should be shown as -6%.

alya commented:
> I am not sure whether we need tooltips in settings UI as we show the current value there. 

I'm currently feeling OK about leaving them; we could consider again once the comment above has been addressed.

alya commented:
> Am not sure about the colors in disabled state, so would be good to have some feedback on it. 

@terpimost 's approach is to handle that with reduced opacity. I'm not sure what the right number is or what we do elsewehre, maybe something like 0.3 or 0.4?

sahil839 commented:
> The % numbers are wrong. They should all be %s of 140. So, e.g., 131/140 ~= 0.935, so it should be shown as -6%.

#33186 mentions that we should have "-18%" for 122%, so -9% for 140% and I feel that's correct because 140 value is for `web_line_height_percent`.

alya commented:
Oh, I see. Hmm.

alya commented:
OK, chatting with @timabbott , let's try simplifying the whole %s thing (which wasn't super working for me when testing). Let's display it as:

140 -> 0
other numbers -> number of steps in the UI from 140. So, e.g., 131 -> -1, 122 -> -2.

If someone changes it in the API, we can show their current steps value rounded to 1 digit past the decimal (e.g., -1.2).

alya commented:
Follow-up help center PR, which could be appended over here if that works better: #33848.

timabbott commented on web/src/server_events_dispatch.js:
```diff
@@ -904,19 +904,6 @@ export function dispatch_normal_event(event) {
-            ) {
-                information_density.set_base_typography_css_variables();
-                information_density.calculate_timestamp_widths();
-            }
```
Let's replace this with an explicit noop block, rather than deleting it.

sahil839 replied:
There was actually a mistake here where the value in `user_settings` was being updated without this functions be called. So, fixed that and added a comment as well.

timabbott commented on web/tests/dispatch.test.cjs:
```diff
@@ -1040,29 +1040,6 @@ run_test("user_settings", ({override}) => {
-    override(user_settings, "web_font_size_px", 122);
-    override(information_density, "set_base_typography_css_variables", noop);
-    dispatch(event);
-    assert_same(user_settings.web_line_height_percent, 130);
```
Similarly, I think we can keep tests here, just asserting that the values didn't change, with comments.

timabbott commented on web/src/personal_menu_popover.ts:
```diff
@@ -105,6 +106,36 @@ export function initialize(): void {
+            const resizeObserver = new ResizeObserver(() => {
+                void instance.popperInstance?.update();
+            });
+            resizeObserver.observe(document.querySelector("#personal-menu-dropdown")!);
```
Probably we want to request an animation frame a la 196f4e92a3c4e722cdf14379eba3190b1aafdfaa, to avoid that potential failure mode.

sahil839 replied:
Done.

timabbott commented on web/src/information_density.ts:
```diff
@@ -171,3 +198,122 @@ export function initialize(): void {
+        changed_settings.dense_mode = false;
+        $("body").toggleClass("less-dense-mode");
+        $("body").toggleClass("more-dense-mode");
+    }
```
Remind me what `dense_mode` does mechanically? This seems like it drops dense mode, but will never restore it, and I'm not sure if that's intended.

sahil839 replied:
This was incorrect. Updated to correctly add and remove the required classes.

Also, currently we only change `dense_mode` if and we set it to False -
- `dense_mode` is set to `True` originally.
- New values of font size is not 14 and of line height is not 122.

But we do not set it to `True` again if font size is 14 and line height is 122. This is same as was the behavior previously.

timabbott commented on web/styles/popovers.css:
```diff
@@ -1272,6 +1272,17 @@ ul.popover-group-menu-member-list {
+    .info-density-controls {
+        display: flex;
+        padding: 2px 10px;
+        gap: 8px;
```
Probably we should be writing new `gap`/`padding` values in `em`.

timabbott commented on web/templates/popovers/navbar/navbar_personal_menu_popover.hbs:
```diff
@@ -115,6 +115,34 @@
+                            </button>
+                            <button class="info-density-button increase-button">
+                                <i class="zulip-icon zulip-icon-plus" aria-hidden="true"></i>
+                            </button>
```
I think we need `aria-label` and tooltips on these increase/decrease buttons for accessibility.

sahil839 replied:
We have tooltips already that are added in `tippyjs.ts`. And for `aria-label`, can we show the values like we show in tooltips dynamically? Currently, I have set it to texts like "Increase font size", "Reset font size to default", etc.

timabbott commented:
I think the other thing I'm noticing in the personal menu context is that especially the line height ones lack context -- I think we want some version of "Increase line spacing" and "Decrease line spacing" as the first line. Font size may be obvious it's not necessary, but we'll want to consider that once it's updated to better match the issue as discussed above.

alya commented:
Let's not fiddle with the first line yet, but make the second lines more descriptive instead.

_Current font size: 14 pt
Current font size: default (16pt)
Current line spacing: -1
Current line spacing: default_

sahil839 commented:
Updated the PR.
- Updated tooltips to show the 0, 1, -1 values for line height.
- Current value text in tooltips is in italics.
- Also, made the tooltip descriptive as mentioned above.
- Tooltips is shown for disabled buttons and normal cursor is shown when default button is disabled.

sahil839 commented:
Regarding showing buttons as disabled, do we want to reduce the opacity of icons only or we also want to change the opacity of background color. I have not done any changes yet.

sahil839 commented:
And am not sure about how to handle cases where users set the values for these settings other than what we allow in UI using the API.
- Like for font size, user can set the values to 10px and 23px. Of course we can disable and enable the buttons accordingly but what do we want to show on tooltips for those buttons? Like we show "Already at maximum (20)" when set to 20px, so do we want to show it like "Already exceeds the maximum (20)"?
- For line heights also we need to decide the above cases for how to handle tooltips. Apart from tooltips, there are a couple of other things we need to handle.
Like suppose a user sets the value to 152 using the API. That will be displayed 1.3, and then if I press the "+" button, what should be the increment to the value? Normal increment of 9 would exceed the maximum allowed value or should we just increase it to 158 which is allowed.
And similary, when pressing "+" when setting is set to say 133, should it be changed to next standard value, i.e. 140 or it should just be increased by 9, i.e. 142.

alya commented:
> Regarding showing buttons as disabled, do we want to reduce the opacity of icons only or we also want to change the opacity of background color? 

I asked @terpimost on CZO.

alya commented:
> What will be displayed 1.3, and then if I press the "+" button, what should be the increment to the value? Normal increment of 9 would exceed the maximum allowed value or should we just increase it to 158 which is allowed.

The +/- button should take you to the nearest integer value in the appropriate direction (e.g., 1.3 -> 2 or 1.3 -> 1, depending which way you go.)

alya commented:
I think we need to do more fiddling with the tooltips to make it clear what's happening.

## Tooltips for font size buttons

Line 2 is in italics.

| Button  | Disabled?             | Tooltip line 1 | Tooltip line 2 |
|---------|-----------------------|----------------|----------------|
| Minus   | no                    |    Change to  {x}pt  font       |       -         |
| Minus   | yes                   |     Already at minimum font size (12pt)           |                |
| Minus   | yes, outside of range |    Already below recommended minimum font size (12pt)            |                |
| Plus    | no                    |       Change to  {x}pt  font         |                |
| Plus    | yes                   |     Already at maximum font size (20pt)            |                |
| Plus    | yes, outside of range |     Already above recommended maximum font size (20pt)           |                |
| Default | no                    |      Reset to default font size (16pt)          |       Current font size: {x}pt         |
| Default | yes                   |         Already at default font size (16pt)    |                |

alya commented:
## Tooltips for line spacing buttons

Line 2 is in italics.

- Note that it's "+1", not "1" for positive numbers.
- 0 is always written as "default", not the number.
- We don't mention numbers when we tell you it's max/min/default, since they aren't that meaningful.

| Button  | Disabled?             | Tooltip line 1 | Tooltip line 2 |
|---------|-----------------------|----------------|----------------|
| Minus   | no                    |    Change to  {- or +}{x}  line spacing / Change to default line spacing      |               |
| Minus   | yes                   |     Already at minimum line spacing          |                |
| Minus   | yes, outside of range |    Already below recommended minimum line spacing            |                |
| Plus    | no                    |       Change to  {- or +}{x}  line spacing / Change to default line spacing         |                |
| Plus    | yes                   |     Already at maximum line spacing            |                |
| Plus    | yes, outside of range |     Already above recommended maximum line spacing           |                |
| Default | no                    |      Reset to default line spacing          |       Current line spacing: {- or +}{x}         |
| Default | yes                   |         Already at default line spacing    |                |

alya commented:
The cursor isn't quite what I was looking for.

## Cursor, for all buttons
- A hand when not disabled
- Default cursor when disabled (I see a hand in some places and a disabled cursor sometimes when I test.)
We aren't using a disabled cursor at all.

alya commented:
For a non-admin, the default user setting doesn't look great visually -- it should all look disabled:

![Screenshot 2025-03-06 at 10 04 55](https://github.com/user-attachments/assets/47d55c97-268f-4939-9949-9d577437339b)

We should also turn off all the tooltips there for non-admins -- currently, it tells me I'm at max and min and default, depending on where I hover. :)

We also don't properly manage the UI for disabling the theme setting, but we can file a separate issue for that.

sahil839 commented:
> > What will be displayed 1.3, and then if I press the "+" button, what should be the increment to the value? Normal increment of 9 would exceed the maximum allowed value or should we just increase it to 158 which is allowed.
> 
> The +/- button should take you to the nearest integer value in the appropriate direction (e.g., 1.3 -> 2 or 1.3 -> 1, depending which way you go.)

And if suppose the value is set to greater than recommended maximum like 4.2, or 4. Then clicking on "-" button should take it to 3 which is also one of the non-recommended values or to 2. And similar case will be there for font size except there will only be integer values for that, like clicking "-" when font size is set to 24pt should change the setting to 23pt or 20pt.

alya commented:
If you're outside the range, clicking on the button should take you to the closest recommended value, so 2 or 20pt in your examples.

sahil839 commented:
Updated the PR.
- Updated the tooltips. No tooltips are shown to non-admins in "Default user settings" panel.
- Updated the behavior when value of settings is set to other than the recommended values.
- Added a commit to reduce width of theme switcher in spectators menu and also commit to add UI for font size and line height settings.
- Fixed the disabled state in default user settings panel.

sahil839 commented:
Having some trouble aligning the text for smaller font sizes like 12px and 13px. @karlstolley any ideas on what is wrong?

![image](https://github.com/user-attachments/assets/228d6b3d-e358-4318-b0d7-4f74268c3bce)

alya commented:
Looks like this still needs to be addressed:

> Note that it's "+1", not "1" for positive numbers.

alya commented:
I'm still feeling like disabled buttons are a bit too close to non-disabled. What if we just tried changing to 0.4 opacity?

alya commented:
In spectator mode, the theme picker seems a bit too narrow. I don't think it's edges are quite lining up with the font size picker and the text options.

I would also like to see more vertical spacing between the two widgets; it feels a bit crowded.

![Screenshot 2025-03-10 at 09 21 46@2x](https://github.com/user-attachments/assets/8d521472-a632-4846-a14d-49c55ea49f15)

alya commented:
The rest looks good to me. I didn't test the behavior of setting the config outside the recommended bounds.

sahil839 commented:
- Updated line spacing to show "1" as "+1" both in tooltips and in the settings where current value is shown.
- Fixed the spectator modal
- Increased the tooltip delay from 425ms to 750ms.
- For disabled buttons, changed the opacity to 0.4 for now and have shared screenshots on CZO.

![spectator-menu](https://github.com/user-attachments/assets/790f83fd-3e91-4aec-9018-e0d0e85e8527)

timabbott commented on web/src/information_density.ts:
```diff
@@ -171,3 +200,259 @@ export function initialize(): void {
+    }
+}
+
+const line_height_supported_values = [122, 131, 140, 149, 158];
```
This can be a follow-up, but this should probably be computed from `INFO_DENSITY_VALUES_DICT`

timabbott replied:
Added a TODO comment, and moved this to be next to `INFO_DENSITY_VALUES_DICT`

timabbott commented on web/src/information_density.ts:
```diff
@@ -171,3 +200,259 @@ export function initialize(): void {
+        return INFO_DENSITY_VALUES_DICT[property].minimum;
+    }
+
+    assert(property === "web_line_height_percent");
```
I guess it'd be nice to have a comment here, noting that we only support integer font sizes, but there are valid line height values in between the steps here.

timabbott replied:
``` diff
@@ -242,6 +243,10 @@ export function find_new_supported_value_for_setting(
         return INFO_DENSITY_VALUES_DICT[property].minimum;
     }
 
+    // We know the value is inside the range of valid values, but not
+    // a recommended value. This is only possible with line height,
+    // where we allow any integer in the database, but only offer
+    // certain steps in the UI.
     assert(property === "web_line_height_percent");
 
     if ($elem.hasClass("increase-button")) {
```

timabbott commented on web/src/information_density.ts:
```diff
@@ -171,3 +200,259 @@ export function initialize(): void {
+                );
+            }
+        } else if (!$elem.prop("disabled")) {
+            tooltip_first_line = $t({defaultMessage: "Change to {new_value}pt font"}, {new_value});
```
I think "font size" rather than "font" is expected for this string?

timabbott replied:
``` diff
diff --git a/web/src/information_density.ts b/web/src/information_density.ts
index 0eb10bfc92..3d985a2519 100644
--- a/web/src/information_density.ts
+++ b/web/src/information_density.ts
@@ -368,7 +368,7 @@ export function get_tooltip_context_for_info_density_buttons(
                 );
             }
         } else if (!$elem.prop("disabled")) {
-            tooltip_first_line = $t({defaultMessage: "Change to {new_value}pt font"}, {new_value});
+            tooltip_first_line = $t({defaultMessage: "Change to {new_value}pt font size"}, {new_value});
         } else {
             if ($elem.hasClass("increase-button")) {
                 const maximum_value = INFO_DENSITY_VALUES_DICT[property].maximum;
```

timabbott commented on web/src/personal_menu_popover.ts:
```diff
@@ -105,6 +106,34 @@ export function initialize(): void {
+                    url: "/json/settings",
+                    data,
+                });
+                e.preventDefault();
```
We probably want a comment here explaining the offline reasoning behind not having `success` and `error` handlers?

timabbott replied:
``` diff
diff --git a/web/src/personal_menu_popover.ts b/web/src/personal_menu_popover.ts
index e4e90bc433..2789060da5 100644
--- a/web/src/personal_menu_popover.ts
+++ b/web/src/personal_menu_popover.ts
@@ -122,6 +122,20 @@ export function initialize(): void {
                 void channel.patch({
                     url: "/json/settings",
                     data,
+                    // We don't declare success or error
+                    // handlers. We've already locally echoed the
+                    // change, and the thinking for this component is
+                    // that right answer for error handling is to do
+                    // nothing. For the offline case, just letting you
+                    // adjust the font size locally is great, and it's
+                    // not obvious what good error handling is here
+                    // for the server being down other than "try again
+                    // later", which might as well be your next
+                    // session.
+                    //
+                    // This strategy also avoids unpleasant races
+                    // involving the button being clicked several
+                    // times in quick succession.
                 });
                 e.preventDefault();
             });
```

timabbott commented on web/src/information_density.ts:
```diff
@@ -171,3 +200,259 @@ export function initialize(): void {
+}
+
+export function get_string_display_value_for_line_height(setting_value: number): string {
+    let display_value = (setting_value - NON_COMPACT_MODE_LINE_HEIGHT_PERCENT) / 9;
```
``` diff
diff --git a/web/src/information_density.ts b/web/src/information_density.ts
index 85b6c80d16..1dc855a4d8 100644
--- a/web/src/information_density.ts
+++ b/web/src/information_density.ts
@@ -306,7 +306,9 @@ export function update_information_density_settings(
 }
 
 export function get_string_display_value_for_line_height(setting_value: number): string {
-    let display_value = (setting_value - NON_COMPACT_MODE_LINE_HEIGHT_PERCENT) / 9;
+    let display_value =
+        (setting_value - NON_COMPACT_MODE_LINE_HEIGHT_PERCENT) /
+        INFO_DENSITY_VALUES_DICT.web_line_height_percent.step_value;
 
     if (display_value % 1 === 0) {
         // If value is an integer, we just return here to avoid showing
```

timabbott commented:
I made the small tweaks above and also fixed the linter errors as follows:

``` diff
diff --git a/web/src/information_density.ts b/web/src/information_density.ts
index e4c37f438c..323d684812 100644
--- a/web/src/information_density.ts
+++ b/web/src/information_density.ts
@@ -311,19 +311,20 @@ export function update_information_density_settings(
 }
 
 export function get_string_display_value_for_line_height(setting_value: number): string {
-    let display_value =
+    const step_count =
         (setting_value - NON_COMPACT_MODE_LINE_HEIGHT_PERCENT) /
         INFO_DENSITY_VALUES_DICT.web_line_height_percent.step_value;
+    let display_value;
 
-    if (display_value % 1 === 0) {
+    if (step_count % 1 === 0) {
         // If value is an integer, we just return here to avoid showing
         // 1.0 for 1.
-        display_value = display_value.toString();
+        display_value = step_count.toString();
     } else {
-        display_value = display_value.toFixed(1);
+        display_value = step_count.toFixed(1);
     }
 
-    if (display_value > 0) {
+    if (step_count > 0) {
         // We want to show "1" as "+1".
         return "+" + display_value;
     }
```

timabbott commented:
With those changes, this looks great, merged, thanks @sahil839! 

I think we're most likely to find additional changes we want to make to the UI after this reaches chat.zulip.org.
